drivers/dma/dma_nxp_edma.c

```diff
@@ -0,0 +1,669 @@
+		LOG_ERR("channel ID %u is not valid", chan_id);
+		return -EINVAL;
+	}
+
+	prev_state = chan->state;
+
+	/* change channel's state to STOPPED */
+	ret = channel_change_state(chan, CHAN_STATE_STOPPED);
+	if (ret < 0) {
+		LOG_ERR("failed to change channel %d state to STOPPED", chan_id);
+		return ret;
+	}
+
+	LOG_DBG("stopping channel %u", chan_id);
+
+	if (prev_state == CHAN_STATE_SUSPENDED) {
+		/* if the channel has been suspended then there's
+		 * no point in disabling the HW requests again. Just
+		 * jump to the channel release operation.
+		 */
+		goto out_release_channel;
+	}
+
+	/* disable HW requests */
+	EDMA_ChannelRegUpdate(data->hal_cfg, chan_id, EDMA_TCD_CH_CSR, 0,
+			      EDMA_TCD_CH_CSR_ERQ_MASK);
+out_release_channel:
+
+	/* clear the channel MUX so that it can used by a different peripheral.
+	 *
+	 * note: because the channel is released during dma_stop() that means
+	 * dma_start() can no longer be immediately called. This is because
+	 * one needs to re-configure the channel MUX which can only be done
+	 * through dma_config(). As such, if one intends to reuse the current
+	 * configuration then please call dma_suspend() instead of dma_stop().
+	 */
+	if (EDMA_HAS_MUX(data->hal_cfg)) {
+		ret = EDMA_SetChannelMux(data->hal_cfg, chan_id, 0);
+		if (ret < 0) {
+			LOG_ERR("failed to set channel MUX");
+			return to_std_error(ret);
+		}
+	}
+
+	edma_dump_channel_registers(data, chan_id);
+
+	return 0;
+}
+
+static int edma_start(const struct device *dev, uint32_t chan_id)
+{
+	struct edma_data *data;
+	const struct edma_config *cfg;
+	struct edma_channel *chan;
+	int ret;
+
+	data = dev->data;
+	cfg = dev->config;
+
+	/* fetch channel data */
+	chan = lookup_channel(dev, chan_id);
+	if (!chan) {
+		LOG_ERR("channel ID %u is not valid", chan_id);
+		return -EINVAL;
+	}
+
+	/* change channel's state to STARTED */
+	ret = channel_change_state(chan, CHAN_STATE_STARTED);
+	if (ret < 0) {
+		LOG_ERR("failed to change channel %d state to STARTED", chan_id);
+		return ret;
+	}
+
+	LOG_DBG("starting channel %u", chan_id);
+
+	/* enable HW requests */
+	EDMA_ChannelRegUpdate(data->hal_cfg, chan_id,
+			      EDMA_TCD_CH_CSR, EDMA_TCD_CH_CSR_ERQ_MASK, 0);
+
+	return 0;
+}
+
+static int edma_reload(const struct device *dev, uint32_t chan_id, uint32_t src,
+		       uint32_t dst, size_t size)
+{
+	struct edma_data *data;
+	struct edma_channel *chan;
+	int ret;
+	unsigned int key;
+
+	data = dev->data;
+
+	/* fetch channel data */
+	chan = lookup_channel(dev, chan_id);
+	if (!chan) {
+		LOG_ERR("channel ID %u is not valid", chan_id);
+		return -EINVAL;
+	}
+
+	/* channel needs to be started to allow reloading */
+	if (chan->state != CHAN_STATE_STARTED) {
+		LOG_ERR("reload is only supported on started channels");
+		return -EINVAL;
+	}
+
+	if (chan->cyclic_buffer) {
+		key = irq_lock();
+		ret = EDMA_CHAN_PRODUCE_CONSUME_B(chan, size);
+		irq_unlock(key);
+		if (ret < 0) {
+			LOG_ERR("chan %d buffer overflow/underrun", chan_id);
```

Comment:
I see that here you return in case of error.
On EDMA_CHAN_PRODUCE_CONSUME_A() you don't.

Reply:
As opposed to the ISR, here we can actually let the user know something went wrong. There's also no callback here which should (arguably) be called even if something bad happened.